ReSwift/CoreTypes/Store.swift

```diff
@@ -91,61 +83,61 @@ public class Store<State: StateType>: StoreType {
+        (_ subscriber: S, selector: ((State) -> SelectedState)?) {
+            if !_isNewSubscriber(subscriber: subscriber) { return }
+
+            subscriptions.append(Subscription(subscriber: subscriber, selector: selector))
+
+            if let state = self.state {
+                subscriber._newState(state: selector?(state) ?? state)
+            }
+    }
+    #else
     public func subscribe<SelectedState, S: StoreSubscriber
         where S.StoreSubscriberStateType == SelectedState>
         (subscriber: S, selector: ((State) -> SelectedState)?) {
-            #if swift(>=3)
-                if !_isNewSubscriber(subscriber: subscriber) { return }
-            #else
-                if !_isNewSubscriber(subscriber) { return }
-            #endif
+            if !_isNewSubscriber(subscriber) { return }
 
             subscriptions.append(Subscription(subscriber: subscriber, selector: selector))
 
             if let state = self.state {
-                #if swift(>=3)
-                    subscriber._newState(state: selector?(state) ?? state)
-                #else
-                    subscriber._newState(selector?(state) ?? state)
-                #endif
+                subscriber._newState(selector?(state) ?? state)
             }
     }
+    #endif
 
+    #if swift(>=3)
+    public func unsubscribe(_ subscriber: AnyStoreSubscriber) {
+        if let index = subscriptions.index(where: { return $0.subscriber === subscriber }) {
+            subscriptions.remove(at: index)
+        }
+    }
+    #else
     public func unsubscribe(subscriber: AnyStoreSubscriber) {
-        #if swift(>=3)
-            if let index = subscriptions.index(where: { return $0.subscriber === subscriber }) {
-                subscriptions.remove(at: index)
-            }
-        #else
-            if let index = subscriptions.indexOf({ return $0.subscriber === subscriber }) {
-                subscriptions.removeAtIndex(index)
-            }
-        #endif
+        if let index = subscriptions.indexOf({ return $0.subscriber === subscriber }) {
+            subscriptions.removeAtIndex(index)
+        }
     }
+    #endif
 
     public func _defaultDispatch(action: Action) -> Any {
-        if isDispatching {
-            // Use Obj-C exception since throwing of exceptions can be verified through tests
-            #if swift(>=3)
-                NSException.raise(
-                    "ReSwift:IllegalDispatchFromReducer" as NSExceptionName,
-                    format: "Reducers may not dispatch actions.",
-                    arguments: getVaList(["nil"]))
-            #else
-                NSException.raise(
-                    "ReSwift:IllegalDispatchFromReducer",
-                    format: "Reducers may not dispatch actions.",
-                    arguments: getVaList(["nil"]))
-            #endif
+        guard !isDispatching else {
+            fatalError("ReSwift:IllegalDispatchFromReducer - Reducers may not dispatch actions.")
         }
```

Comment:
So I remembered what the issue was: `XCTAssertThrowsSpecific` is not available in Swift 2.

We either need to:

- Include the Nimble Objective C shims for catching NSExceptions: https://github.com/Quick/Nimble/blob/master/Sources/Nimble/Adapters/ObjectiveC/NMBExceptionCapture.m
    - Still have to use NSException

- Use a posix signal catcher: http://www.cocoawithlove.com/blog/2016/02/02/partial-functions-part-two-catching-precondition-failures.html
    - Catches NSException or fatalErrors

- Move the refactor the Assertion shim to be internal only.
    - Only catches fatalError

At the moment I have renamed the shimmed `fatalError` to `raiseFatalError` (just to be absolute clear) and marked them all as internal, which still works in testing as long as the framework is `@testable import ReSwift`. I'll commit it after moving the comments.

Happy to look at either of the alternatives or another suggestion.

We COULD make `_defaultDispatch` throw, but that would require refactoring every app that calls `dispatch`.